file_flattener.go

```diff
@@ -0,0 +1,275 @@
+// Licensed to The Moov Authors under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. The Moov Authors licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package ach
+
+import (
+	"errors"
+	"fmt"
+	"sort"
+)
+
+var (
+	ErrFlattenChangedEntryCount   = errors.New("Flatten operation changed entry and addenda count")
+	ErrFlattenChangedDebitAmount  = errors.New("Flatten operation changed total debit entry amount")
+	ErrFlattenChangedCreditAmount = errors.New("Flatten operation changed total credit entry amount")
+)
+
+// FlattenFile returns a flattened version of a File, where batches with similar batch headers are consolidated.
+//
+// Two batches are eligible to be combined if:
+//   - their headers match, excluding the batch number (which isn't used in return matching and reflects
+//     the final composition of the file.)
+//   - they don't contain any entries with common trace numbers, since trace numbers must be unique
+//     within a batch.
+func FlattenFile(originalFile *File) (*File, error) {
```

Comment:
nit: This name (and `FlattenBatches`) make it sound a little like we're mutating the `File` rather than returning a flattened version. I actually thought `FlattenBatches` wasn't working when I first used it. Maybe it's just me though, not that important.

Reply:
Would `Flatten(file) (*File, error)` make more sense? You're right it sounds like a mutation operation.

Reply:
`Flatten` is a bit better I think. `Flattened` or `FlattenedFile` feel the most to me like we are producing a new object, but I feel less conviction about it the more I think about it; this is a hard naming problem 😬 cool with whatever.